File: HW1/javaApp/src/main/java/io/rienel/domain/Panther.java
package io.rienel.domain;

import java.util.Objects;

/**
 * todo armotozov
 *
 * @author Artem Motozov
 * @since 2022.09.19
 */
public class Panther extends Animal<Animal>{

    private final Boolean isBlack;

    public Panther(final String name,
                   final Double weight,
                   final Double fullHeight,
                   final Double fullLength,
                   final Double fullWidth) {
        super(name, weight, AnimalType.PANTHER, fullHeight, fullLength, fullWidth);
        isBlack = true;
    }

    public Panther(final String name,
                   final Double weight,
                   final Double fullHeight,
                   final Double fullLength,
                   final Double fullWidth,
                   final Boolean isBlack) {
        super(name, weight, AnimalType.PANTHER, fullHeight, fullLength, fullWidth);
        this.isBlack = isBlack;
    }

    public Boolean getBlack() {
        return isBlack;
    }

    @Override
    public void setName(String name) {
        Objects.requireNonNull(name);
        super.setName(name);
    }

    @Override
    public void setWeight(Double weight) {
        Objects.requireNonNull(weight);
        if (weight < 10.0 || weight > 50.0) {
            throw new IllegalArgumentException("Illegal weight");
        }
        super.setWeight(weight);
    }

    @Override
    public void setFullHeight(Double fullHeight) {
        Objects.requireNonNull(fullHeight);
        if (fullHeight < 2.0 || fullHeight > 10.0) {
            throw new IllegalArgumentException("Illegal fullHeight");
        }
        super.setFullHeight(fullHeight);
    }

    @Override
    public void setFullLength(Double fullLength) {
        Objects.requireNonNull(fullLength);
        if (fullLength < 2.0 || fullLength > 10.0) {
            throw new IllegalArgumentException("Illegal fullLength");
        }
        super.setFullLength(fullLength);
    }

    @Override
    public void setFullWidth(Double fullWidth) {
        Objects.requireNonNull(fullWidth);
        if (fullWidth < 2.0 || fullWidth > 10.0) {
            throw new IllegalArgumentException("Illegal fullWidth");
        }
        super.setFullWidth(fullWidth);
    }

    @Override
    public Boolean eat(Animal food) {
        Objects.requireNonNull(food);

        System.out.println("PANTHER: EAT ALL");
        return true;
    }

    @Override
    public String speak() {
        if (isBlack) {
            return "PANTHER: ВАКАНДА ФОРЕВЕР";
        } else {
            return "PANTHER: ROARMEOW";
        }
    }
}

File: HW1/kotlinApp/src/main/kotlin/io/rienel/domain/Panther.kt
package io.rienel.domain

import io.rienel.domain.interfaces.Feedable
import io.rienel.domain.interfaces.Speekable

/**
 * todo armotozov
 *
 * @author Artem Motozov
 * @since 2022.09.19
 */
class Panther(
    name: String,
    weight: Double,
    fullHeight: Double,
    fullLength: Double,
    fullWidth: Double,
    private var isBlack: Boolean
) : Animal(
    name,
    weight,
    AnimalType.PANTHER,
    fullHeight,
    fullLength,
    fullWidth
), Feedable<Animal>, Speekable {

    override var name: String
        get() = super.name
        set(value) {
            assert(value.isNotEmpty())
            super.name = value
        }
    override var weight: Double
        get() = super.weight
        set(value) {
            assert(value in 20.0 .. 200.0)
            super.weight = value
        }

    override var fullHeight: Double
        get() = super.fullHeight
        set(value) {
            assert(fullHeight in 50.0 .. 100.0)
            super.fullHeight = value
        }

    override var fullLength: Double
        get() = super.fullLength
        set(value) {
            assert(fullLength in 50.0 .. 100.0)
            super.fullLength = value
        }
    override var fullWidth: Double
        get() = super.fullWidth
        set(value) {
            assert(fullWidth in 50.0 .. 100.0)
            super.fullWidth = value
        }

    override fun eat(food: Animal): Boolean {
        println("PANTHER: EAT ALL ANIMALS")
        return true
    }

    override fun speak(): String {
        return if (isBlack) {
            "ВАКАНДА ФОРЕВЕР"
        } else {
            "PANTHER: ROARMEOW"
        }
    }
}